diff --git a/drivers/gpu/drm/panel/panel-simple.c b/drivers/gpu/drm/panel/panel-simple.c
index 4eb0e14b1..606e0239f 100644
--- a/drivers/gpu/drm/panel/panel-simple.c
+++ b/drivers/gpu/drm/panel/panel-simple.c
@@ -1396,12 +1396,51 @@ static const struct panel_desc foxlink_fl500wvr00_a0t = {
 	.num_modes = 1,
 	.bpc = 8,
 	.size = {
-		.width = 108,
+		.width = 37,
 		.height = 65,
 	},
 	.bus_format = MEDIA_BUS_FMT_RGB888_1X24,
 };
 
+// static const struct drm_display_mode lattland_mostima_mode = {
+// 	.clock = 12000,
+// 	.hdisplay = 480,
+// 	.hsync_start = 480 + 24,
+// 	.hsync_end = 480 + 24 + 6,
+// 	.htotal = 480 + 24 + 6 + 18,
+// 	.vdisplay = 640,
+// 	.vsync_start = 640 + 16,
+// 	.vsync_end = 640 + 16 + 4,
+// 	.vtotal = 640 + 16 + 4 + 10,
+// 	.vrefresh = 60,
+// };
+static const struct drm_display_mode lattland_mostima_mode = {
+	.clock = 24000,
+	.hdisplay = 384,
+	.hsync_start = 384 + (24 + 60 - 24),    // extra 24 pixel for alignment
+	.hsync_end = 384 + (24 + 60 - 24) + 6,
+	.htotal = 384 + (24 + 60 - 24) + 6 + (18 + 60),
+	.vdisplay = 640,
+	.vsync_start = 640 + 16,
+	.vsync_end = 640 + 16 + 4,
+	.vtotal = 640 + 16 + 4 + 10 - 1,
+	.vrefresh = 60,
+};
+
+static const struct panel_desc lattland_mostima = {
+	.modes = &lattland_mostima_mode,
+	.num_modes = 1,
+	// .bpc = 8,
+	.bpc = 6,
+	.size = {
+		.width = 108,
+		.height = 65,
+	},
+	.bus_format = MEDIA_BUS_FMT_RGB565_1X16,
+	// .bus_format = MEDIA_BUS_FMT_RGB888_1X24,
+	// .bus_flags = DRM_BUS_FLAG_PIXDATA_NEGEDGE,
+};
+
 static const struct drm_display_mode friendlyarm_hd702e_mode = {
 	.clock		= 67185,
 	.hdisplay	= 800,
@@ -3271,6 +3310,9 @@ static const struct of_device_id platform_of_match[] = {
 	}, {
 		.compatible = "foxlink,fl500wvr00-a0t",
 		.data = &foxlink_fl500wvr00_a0t,
+	}, {
+		.compatible = "lattland,mostima",
+		.data = &lattland_mostima,
 	}, {
 		.compatible = "friendlyarm,hd702e",
 		.data = &friendlyarm_hd702e,
